std/selector/slice.go

```diff
@@ -0,0 +1,87 @@
+package selector
+
+import (
+	"github.com/consensys/gnark/constraint/solver"
+	"github.com/consensys/gnark/frontend"
+	"math/big"
+)
+
+func init() {
+	// register hints
+	solver.RegisterHint(StepOutput)
+}
+
+// Partition selects left or right side of the input array, with respect to the pivotPosition.
+// More precisely when rightSide is false, for each i we have:
+//
+//	if i < pivotPosition
+//	    out[i] = input[i]
+//	else
+//	    out[i] = 0
+//
+// and when rightSide is true, for each i we have:
+//
+//	if i >= pivotPosition
+//	    out[i] = input[i]
+//	else
+//	    out[i] = 0
+//
+// We must have pivotPosition >= 1 and pivotPosition <= len(input)-1, otherwise a proof cannot be generated.
+func Partition(api frontend.API, pivotPosition frontend.Variable, rightSide bool,
+	input []frontend.Variable) (out []frontend.Variable) {
+	out = make([]frontend.Variable, len(input))
+	var mask []frontend.Variable
+	// we create a bit mask to multiply with the input
+	if rightSide {
+		mask = StepMask(api, len(input), pivotPosition, 0, 1)
+	} else {
+		mask = StepMask(api, len(input), pivotPosition, 1, 0)
+	}
+	for i := 0; i < len(out); i++ {
+		out[i] = api.Mul(mask[i], input[i])
+	}
+	return
+}
+
+// StepMask generates a step like function into an output array of a given length.
+// The output is an array of length outputLen,
+// such that its first stepPosition elements are equal to startValue and the remaining elements are equal to
+// endValue. Note that outputLen cannot be a circuit variable.
+//
+// We must have stepPosition >= 1 and stepPosition <= outputLen, otherwise a proof cannot be generated.
+// This function panics when outputLen is less than 2.
+func StepMask(api frontend.API, outputLen int,
+	stepPosition, startValue, endValue frontend.Variable) []frontend.Variable {
+	if outputLen < 2 {
+		panic("the output len of StepMask must be >= 2")
+	}
+	// get the output as a hint
+	out, _ := api.Compiler().NewHint(StepOutput, outputLen, stepPosition, startValue, endValue)
```

Comment:
Handle error. It should be sufficient to do `panic(fmt.Sprintf("step hint error: %v", err))`. I do not recommend adding error return to the function as it will be more difficult to use and safe recovery is impossible anyway (there is something wrong internally if err happens).

Reply:
Very good point.